src/Npgsql/TypeHandlers/PostgisGeometryHandler.cs

```diff
@@ -0,0 +1,498 @@
+                    return false;
+                _buf.Skip(1); // byte storing the endianness of the data structure (canonical form = NDR).
+                _id = _buf.PostgisReadUInt32();
+            }
+            if (!_srid.HasValue)
+            {
+                if ((_id & (uint)OgrModifier.HasSRID) != 0)
+                {
+                    if (_buf.ReadBytesLeft < 4)
+                        return false;
+                    _srid = _buf.PostgisReadUInt32();
+                }
+                else
+                {
+                    _srid = 0;
+                }
+            }
+
+            switch ((OgrIdentifier)(_id & (uint)7))
+            {
+                case OgrIdentifier.Point:
+                    if (_buf.ReadBytesLeft < 16)
+                        return false;
+                    result = new PostgisPoint(_buf.PostgisReadDouble(), _buf.PostgisReadDouble());
+                    result.SRID = _srid.Value;
+                    return true;
+
+                case OgrIdentifier.LineString:
+                    if (_ipts == -1)
+                    {
+                        if (_buf.ReadBytesLeft < 4)
+                            return false;
+                        _points = new BBPoint[_buf.PostgisReadInt32()];
+                        _ipts = 0;
+                    }
+                    for (; _ipts < _points.Length; _ipts++)
+                    {
+                        if (_buf.ReadBytesLeft < 16)
+                            return false;
+                        _points[_ipts] = new BBPoint(_buf.PostgisReadDouble(), _buf.PostgisReadDouble());
+                    }
+                    result = new PostgisLineString(_points);
+                    result.SRID = _srid.Value;
+                    return true;
+
+                case OgrIdentifier.Polygon:
+                    if (_irng == -1)
+                    {
+                        if (_buf.ReadBytesLeft < 4)
+                            return false;
+                        _rings = new BBPoint[_buf.PostgisReadInt32()][];
+                        _irng = 0;
+                    }
+
+                    for (; _irng < _rings.Length; _irng++)
+                    {
+                        if (_ipts == -1)
+                        {
+                            if (_buf.ReadBytesLeft < 4)
+                                return false;
+                            _rings[_irng] = new BBPoint[_buf.PostgisReadInt32()];
+                            _ipts = 0;
+                        }
+                        for (; _ipts < _rings[_irng].Length; _ipts++)
+                        {
+                            if (_buf.ReadBytesLeft < 16)
+                                return false;
+                            _rings[_irng][_ipts] = new BBPoint(_buf.PostgisReadDouble(), _buf.PostgisReadDouble());
+                        }
+                        _ipts = -1;
+                    }
+                    result = new PostgisPolygon(_rings);
```

Comment:
It seems like a shame to prepare _rings, only to copy it again into the polygon (this is the same as my comment about providing only constructors that copy from IEnumerable inputs).